agent/php_user_instrument_hashmap.c

```diff
@@ -0,0 +1,345 @@
+/*
+ * Copyright 2022 New Relic Corporation. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "php_agent.h"
+#include "php_user_instrument_hashmap.h"
+
+#if ZEND_MODULE_API_NO >= ZEND_7_4_X_API_NO
+
+typedef struct _nr_wraprecs_bucket {
+  struct _nr_wraprecs_bucket* prev;
+  struct _nr_wraprecs_bucket* next;
+  /* for efficiency wraprec is used as both: key and a value */
+  nruserfn_t* wraprec;
+} nr_wraprecs_bucket_t;
+
+typedef struct _nr_php_wraprec_hashmap {
+  nr_php_wraprec_hashmap_dtor_fn_t dtor_func;
+  size_t log2_num_buckets;
+  nr_wraprecs_bucket_t** buckets;
+  size_t elements;
+} nr_php_wraprec_hashmap_t;
+
+static inline size_t nr_count_buckets(const nr_php_wraprec_hashmap_t* hashmap) {
+  return (size_t)(1 << hashmap->log2_num_buckets);
+}
+
+static nr_php_wraprec_hashmap_t* nr_php_wraprec_hashmap_create_internal(
+    size_t log2_num_buckets,
+    nr_php_wraprec_hashmap_dtor_fn_t dtor_fn) {
+  nr_php_wraprec_hashmap_t* hashmap;
+
+  if (0 == log2_num_buckets) {
+    /*
+     * Encode the default value in one place: namely, here.
+     */
+    log2_num_buckets = 8;
+  } else if (log2_num_buckets > 24) {
+    /*
+     * Basic sanity check: it's extremely unlikely that we'll ever need a
+     * hashmap for the user function wraprecs that has more than 2^24 buckets.
+     */
+    log2_num_buckets = 24;
+  }
+
+  hashmap
+      = (nr_php_wraprec_hashmap_t*)nr_malloc(sizeof(nr_php_wraprec_hashmap_t));
+  hashmap->dtor_func = dtor_fn;
+  hashmap->log2_num_buckets = log2_num_buckets;
+  hashmap->buckets = (nr_wraprecs_bucket_t**)nr_calloc(
+      (1 << log2_num_buckets), sizeof(nr_wraprecs_bucket_t*));
+  hashmap->elements = 0;
+
+  return hashmap;
+}
+
+nr_php_wraprec_hashmap_t* nr_php_wraprec_hashmap_create_buckets(
+    size_t buckets,
+    nr_php_wraprec_hashmap_dtor_fn_t dtor_fn) {
+  size_t actual_buckets;
+
+  if (0 == buckets) {
+    actual_buckets = 0;
+  } else {
+    /*
+     * Calculate ceil(log2(buckets)), since that's what
+     * nr_hashmap_create_internal requires.
+     */
+
+    actual_buckets = 1;
+    while (((size_t)(1 << actual_buckets)) < buckets) {
```

Comment:
is there a maximum value that `buckets` should be checked for before entering this loop?

Reply:
The final sanity check for maximum buckets that can ever be in the hashmap is done [here](https://github.com/misiekn/newrelic-php-agent/blob/refactor/wraprec-lookup/agent/php_user_instrument_hashmap.c#L39-L44). The code itself is a copy of [axiom's hashmap implementation](https://github.com/misiekn/newrelic-php-agent/blob/refactor/wraprec-lookup/axiom/util_hashmap.c#L23-L42) which also doesn't have any checks for `buckets` argument.